Extend/Controller/OrderController.php

```diff
@@ -0,0 +1,268 @@
+use \Wirecard\PaymentSdk\Response\Response;
+use \Wirecard\PaymentSdk\Entity\Status;
+
+use \Wirecard\Oxid\Extend\Model\Payment_Gateway;
+use \Wirecard\Oxid\Core\Helper;
+use \Wirecard\Oxid\Extend\Model\Order;
+
+use \Psr\Log\LoggerInterface;
+
+/**
+ * Class Order
+ *
+ * @mixin \OxidEsales\Eshop\Application\Controller\OrderController
+ */
+class OrderController extends OrderController_parent
+{
+    /**
+     * Extends the parent init function and finalizes the order in case it was a Wirecard payment method
+     */
+    public function init()
+    {
+        parent::init();
+
+        $oConfig = Registry::getConfig();
+        $sWdPaymentRedirect = $oConfig->getRequestParameter('wdpayment');
+        $oSession = Registry::getSession();
+        $sWdSessionToken = $oSession->getVariable('wdtoken');
+
+        if ($this->_isPaymentFinished($sWdSessionToken, $sWdPaymentRedirect)) {
+            $sShopBaseUrl = $oConfig->getShopUrl();
+            $sLanguageCode = Registry::getLang()->getBaseLanguage();
+
+            $aParams = array(
+                'lang' => $sLanguageCode,
+                'force_sid' => $oSession->getId(),
+                'stoken' => $oSession->getSessionChallengeToken(),
+                'actcontrol' => 'order',
+                'cl' => 'order',
+                'fnc' => 'execute',
+                'sDeliveryAddressMD5' => $this->getDeliveryAddressMD5(),
+                'challenge' => '',
+                'ord_agb' => '1',
+                'oxdownloadableproductsagreement' => '0',
+                'oxserviceproductsagreement' => '0',
+                'wdtoken' => $sWdSessionToken
+            );
+
+            $sParamStr = http_build_query($aParams);
+
+            $sNewUrl = $sShopBaseUrl . 'index.php?' . $sParamStr;
+            header('Location: '.$sNewUrl);
+        }
+    }
+
+    /**
+     * Redirects to the PaymentSDK if the payment is a Module's method
+     *
+     * @inheritdoc
+     *
+     * @return string
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function execute()
+    {
+        $oBasket = $this->getSession()->getBasket();
+
+        if (!Helper::isModulePaymentMethod($oBasket->getPaymentId())) {
+            return parent::execute();
+        }
+
+        $oOrder = oxNew(Order::class);
+        $sOrderId = Registry::getSession()->getVariable('sess_challenge');
+        $bIsOrderLoaded = $oOrder->load($sOrderId);
+        if ($bIsOrderLoaded) {
+            $iSuccess = $oOrder->oxorder__wdoxidee_finalizeorderstate->value;
+        }
+
+        $oConfig = Registry::getConfig();
+        $sWdPaymentRedirect = $oConfig->getRequestParameter('wdtoken');
+        $sWdSessionToken = $this->getSession()->getVariable('wdtoken');
+        if ($this->_isPaymentFinished($sWdSessionToken, $sWdPaymentRedirect)) {
+            return $this->_getNextStep($iSuccess);
+        }
+
+        if ($bIsOrderLoaded) {
+            $oOrder->delete($sOrderId);
```

Comment:
When does this situation happen? If order already exists is checked in finalizeOrder method. Should we really delete the order? Or somehow log it?

Reply:
If the user clicks on order now and then on the paypal site clicks cancel payment. The old order would stay there forever as pending order because it will never get an associated transaction. We have to add a new order because the user could also add another article to the basket and we cannot update the 'old' already finished order.